### DISCUSSION TITLE
How to: Job encounters error, and should be retried at a later time.

### DISCUSSION THREAD

**Discussion by drusellers** [quartznet/quartznet#2073] [Q&A]
I have a job that is trying to submit an order. Let's say that it runs into an error, the API that its trying to call returns a `429 - Too Many Requests`. I'd like to stop executing the job, and tell it to try again in say 5 minutes.

I'm looking for a pattern on how to best do this.

Job Key `order.submit-order`
Current Trigger Key: `order.<order id>`
Job Data Map: `{ "OrderId": "<order id>" }`
_Assume Job is Idempotent, rerunning N times is just fine_

Option 1: Throw an exception and quartz will stop the job. put it in error state. But I don't see a way to configure how many retries it should go through, and what the delay should be during each one.

Option 2: Given that I have the job key and the trigger key. Additionally, the job has access to the scheduler. When I encounter an error that just needs a little bit of time, I could reschedule the job with the same job key, trigger key (maybe?), and the same Trigger.JobDataMap. Right?

Option 3: Am I missing anything?

TY, I'll add the output of this discussion to a PR for the How To's section.

**Comment by lahma** [comment]
This is always a bit tricky as it has "message queue" characteristics. 

> Let's say that it runs into an error, the API that its trying to call returns a 429 - Too Many Requests. I'd like to stop executing the job, and tell it to try again in say 5 minutes.

I sense that this might also have some global behavior problem as you could have 1000 orders to be handled in Quartz and they might trigger close to each other and you would always create a flood of requests which compete. So I guess your job would have `[DisallowConcurrentExceution]` and orders would be processed one-by-one (probably also durable so job stays in scheduler even if there are no active triggers to handle). 

## Option 1

The good part here is that after some failure you have a definite marker that he precious order was never processed and something needs to be done. But as said, requires some manual intervention.

## Option 2

Self-rescheduling is an option, but that is quite close to having Polly doing the retries inside the job - of course 5 minutes of sleep time inside a job is always long and takes precious worker threads / tasks. 

## Option 3

What about having the trigger run every five minutes (or some other suitable cadence) and after succeeding, unscheduling itself?

```c#
public async Task Execute(IJobExecutionContext context)
{
    if (success)
    {
        await context.Scheduler.UnscheduleJob(context.Trigger.Key);
    }
}
```

**Reply by justmara** [reply]
we use option 4: `Retry` marker attribute on job and IJobListener that listens the JobWasExecuted event and re-schedules job by rules defined in retry-attribute (retry period, fallback delay, retry count, etc) in case if jobException exists :) this way quartz thread is not locked for the whole retry period as it would be in case of polly inside job itself.

[here](https://github.com/quartznet/quartznet/pull/1045) is corresponding discussion, btw

**Comment by jafin** [comment]
more options

## Continuous process loop

Dependency: you are ok with persisting data to a store of some sort.

- Job runs every `<x>` minutes.
- Query the Outstanding_Orders queue store.
- Process any outstanding orders and mark them as processed when successful.
- Additional checks may be needed if jobs are running on the cluster concurrently.

## Job with internal fallback policy

- The job runs as you have it now.
- Job context manages exponential fallback itself with something like Polly
- if after Fallback policy expires then you give up, job is in error state.
- Donwside, jobs remains in process, may be a resource issue if there is a huge backlog.

## Don't use a scheduler
- use a queue/service bus instead.

**Reply by lahma** [reply]
A variant of `Job with internal fallback policy` is to calculate Polly-like next fire time based on invocation count in `catch` phase when error detected and schedule new trigger to run with same trigger data (order id). This well free resources as job isn't locked during retry.

`Don't use a scheduler` is a good thing to always consider, how well does this fit queue processing. Of course you will have same retry policy for messages like in for example MassTransit - if they fail enough many times (like 5) they are moved to error queue and you are again facing manual intervention and re-delivery (which would be re-scheduling in Quartz).
